pkg/controller/perconaservermongodb/pbm.go

```diff
@@ -0,0 +1,503 @@
+			if err != nil {
+				return errors.Wrap(err, "update pitr.compression")
+			}
+		}
+
+		// PBM needs to disabling and enabling PITR to change compression type
+		if err := pbm.SetConfigVar(ctx, "pitr.enabled", "false"); err != nil {
+			return errors.Wrap(err, "disable pitr")
+		}
+		if err := pbm.SetConfigVar(ctx, "pitr.enabled", "true"); err != nil {
+			return errors.Wrap(err, "enable pitr")
+		}
+	}
+
+	val, err = pbm.GetConfigVar(ctx, "pitr.compressionLevel")
+	var compressionLevel *int = nil
+	if err != nil {
+		if errors.Is(err, mongo.ErrNoDocuments) {
+			return nil
+		} else if !errors.Is(err, bsoncore.ErrElementNotFound) {
+			return errors.Wrap(err, "get pitr.compressionLevel")
+		}
+	} else {
+		var iVal int
+		switch v := val.(type) {
+		case int64:
+			iVal = int(v)
+		case int32:
+			iVal = int(v)
+		default:
+			return errors.Errorf("unexpected value of pitr.compressionLevel: %T", val)
+		}
+		compressionLevel = &iVal
+	}
+
+	if !reflect.DeepEqual(compressionLevel, cr.Spec.Backup.PITR.CompressionLevel) {
+		if cr.Spec.Backup.PITR.CompressionLevel == nil {
+			cfg, err := pbm.GetConfig(ctx)
+			if err != nil {
+				return errors.Wrap(err, "get pbm config")
+			}
+
+			cfg.PITR.CompressionLevel = nil
+			if err := pbm.SetConfig(ctx, cfg); err != nil {
+				return errors.Wrap(err, "delete pitr.compressionLevel")
+			}
+		} else {
+			val := strconv.FormatInt(int64(*cr.Spec.Backup.PITR.CompressionLevel), 10)
+			log.Info("Setting pitr.compressionLevel", "value", val)
+			err := pbm.SetConfigVar(ctx, "pitr.compressionLevel", val)
+			if err != nil {
+				return errors.Wrap(err, "update pitr.compressionLevel")
+			}
+		}
+
+		// PBM needs to disabling and enabling PITR to change compression level
+		if err := pbm.SetConfigVar(ctx, "pitr.enabled", "false"); err != nil {
+			return errors.Wrap(err, "disable pitr")
+		}
+		if err := pbm.SetConfigVar(ctx, "pitr.enabled", "true"); err != nil {
+			return errors.Wrap(err, "enable pitr")
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaServerMongoDB) resyncPBMIfNeeded(ctx context.Context, cr *psmdbv1.PerconaServerMongoDB) error {
+	log := logf.FromContext(ctx)
+
+	if cr.Status.State != psmdbv1.AppStateReady || !cr.Spec.Backup.Enabled || !cr.PBMResyncNeeded() {
+		return nil
+	}
+
+	log.Info("resync is needed for all storages")
+
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		c := &psmdbv1.PerconaServerMongoDB{}
+		err := r.client.Get(ctx, types.NamespacedName{Name: cr.Name, Namespace: cr.Namespace}, c)
+		if err != nil {
+			return err
+		}
+
+		orig := c.DeepCopy()
+		delete(c.Annotations, psmdbv1.AnnotationResyncPBM)
+		c.Annotations[psmdbv1.AnnotationResyncInProgress] = "true"
+
+		return r.client.Patch(ctx, c, client.MergeFrom(orig))
+	})
+	if err != nil {
+		return errors.Wrap(err, "delete annotation")
+	}
+
+	// running in separate goroutine to not block reconciliation
+	// until all resync operations finished
+	go func() {
```

Comment:
Here we pass the cr object but it may change in another reconciliation loop, right? I think we should handle this scenario with some form of syncing or queuing i.e. if multiple loops occur if a small amount of time.